pkg/workceptor/remote_work.go

```diff
@@ -68,17 +75,34 @@ func (rw *remoteUnit) connectToRemote(ctx context.Context) (net.Conn, *bufio.Rea
 }
 
 // getConnection retries connectToRemote until connected or the context expires
-func (rw *remoteUnit) getConnection(ctx context.Context) (net.Conn, *bufio.Reader) {
+func (rw *remoteUnit) getConnection(mw *utils.JobContext) (net.Conn, *bufio.Reader) {
 	connectDelay := utils.NewIncrementalDuration(SuccessWorkSleep, MaxWorkSleep, 1.5)
 	for {
-		conn, reader, err := rw.connectToRemote(ctx)
+		conn, reader, err := rw.connectToRemote(mw)
 		if err == nil {
 			return conn, reader
 		}
-		logger.Debug("Connection to %s failed with error: %s",
-			rw.Status().ExtraData.(*remoteExtraData).RemoteNode, err)
+		status := rw.Status()
+		logger.Warning("Connection to %s failed with error: %s",
+			status.ExtraData.(*remoteExtraData).RemoteNode, err)
+		errStr := err.Error()
+		if strings.Contains(errStr, "CRYPTO_ERROR") {
+			detailMsg := "Incorrect tlsclient to remote service"
+			if status.Detail != detailMsg {
```

Comment:
The semantics of UpdateFullStatus are that it gets a lock on the status file, reads its current content, then runs the update function you pass in.  By checking `status.Detail` _outside_ the UpdateFullStatus block, it's possible that someone else was ahead of us in line for the lock, and updated the file before UpdateFullStatus began.  So I would do this as something more like
```
	if strings.Contains(errStr, "CRYPTO_ERROR") {
		var shouldExit = false
		rw.UpdateFullStatus(func(status *StatusFileData) {
			status.Detail = fmt.Sprintf("TLS error connecting to remote service: %s", errStr)
			if status.ExtraData.(*remoteExtraData).RemoteStarted == false {
				shouldExit = true
				status.State = WorkStateFailed
			}
		})
		if shouldExit {
			mw.Cancel()
			return
		}
	}
```

Reply:
To put this another way, everything inside the UpdateFullStatus call is guaranteed(*) to be atomic.

(*) if I wrote the locking code right